### ISSUE TITLE
Duplications insert Object in cloudKit database + solution

### ISSUE THREAD

**Issue by spotlessicode** [mentrena/SyncKit#2]
Hi, 

It's a great Sync method!! First of All really thanks for it! ;)  And I will use for my new IOSApp, and I also wanna create a tutorial for it. I have only one annoying thing with your codes, the duplications after synchronization.  
I have a CoreData stack, and I'm importing for 2 entity 10 and 15 objects at the First launch. Sync is running every time when you open or close the app, but not at First launch (cause problems with import, because it is so slow, that the UI update and other codes can't wait the end of the Sync). And of course, at the very first launch, the user hasn't got any objects, so don't need to sync the database.  
But if you imported these objects, and after that **delete the App, and later download it again** (I'm just test it now from Xcode on real device, and delete the app form device and start running again), the App has a First Launch, so the import starts, but you check just the ObjectID in the CoreDataChangeManager, and not the Object properties. I believe this is a mistake, because, if I create a new Object, the Object ID will be different, even if the object property is the same.  So for 1 entity, there will be not 15 object, there will be 30. **And later when the sync starts, in the App these objects will duplicate**.
 I wrote a code for this situation, to **delete those object, if they have for example the same name property**, or the same displayOrder property. But this is not a perfect solution (I checked the objects at a collection view, but the import is running at launch, sot till i go to that collection view, these objects still duplicated, which can cause other problems too). 
So I suggest a solution, if **you could set for CoreDataChangeManager a "mainattribute"**, wich could set globally in AppDelegate, or anywhere else, as the Synchronizer, than developers could set 1 attribute to every Entity, as a main attribute, which can't be the same, so you should write an if statement for the "mainattribute", in this block: `- (QSSyncedEntity *)createSyncedEntityForRecord:(CKRecord *)record`, to check if it already exists, and if yes, than avoid the targetContext insertation. 

**The second solution this "after deletion" what I'm using now**, if after the Sync, there are duplications, than I delete the objects with the same name or other property. And for those who have Problems with these things **it is strongly recommended to running Sync after every deletion in the App!**  so with this logic:

```
                                      [self.managedObjectContext deleteObject:deleteTarget];
                                      [self.managedObjectContext save:nil];
                                      NSLog(@"deleted!");
                                      [self startSync];
```
And yes there are some words about this SycKit hasn't got Deduplicator, and developer has to write one, but deduplicator for the CoreCata object in the app locally is different as a deduplicator for the CloudKit what you made, so you should improve it. And of course easier modifying your codes and using a global variable implement in a unique App, as with a unique property/database overwrite your SyncKit changemanager, which is otherwise awesome! ;)

**Comment by mentrena** [comment]
Hi there,

Thanks for reporting it.
Yes, using `objectID` seemed like a good idea at first but it's become a limitation --it's also problematic if you perform a lightweight migration of your model, as the `objectID` will likely change.
I was planning to add the option to use a field as primary key/identifier, I think it's the same you're suggesting with that "mainattribute" field. 
Hope to get it done over the Christmas period :)

**Comment by spotlessicode** [comment]
Hi,  Yes! That primary key is that, just I was a little bit confused with the whole structure ;) With that key the whole sync process will be awesome, I didn't find nothing else. We are planning to launch our App at the beginning of the new year. I'll notice you about that! ;) and of course about the tutorial too! So if you are ready, I'm also ready to test it ;) Merry Christmas!

**Comment by mentrena** [comment]
Hi @spotlessicode, I'm working on a fix in this branch:
https://github.com/mentrena/SyncKit/tree/objectIdentifier

So far it seems to be working, but I won't merge just yet as I'd like to make it easy for people (myself) using previous versions of SyncKit to upgrade to 0.3.0. 
Hopefully you'll be able to grab the code from this branch and start using it in your project by making your objects adopt `QSPrimaryKey`.

One other comment that might (or might not) help in your case. In my own app iCloud synchronization can be on or off. When the user turns it on they can choose whether to keep local data or just restore what's on iCloud, which works by throwing away the current database files, creating new empty ones, then using SyncKit to synchronize.
So if I was using devices A and B, delete the app from B and reinstall, when I turn synchronization on in B, then B will create an empty database and synchronize, so in the end B will be an exact copy of A (through iCloud).

**Comment by spotlessicode** [comment]
Thanks! I’ll check it today-tomorrow, as I have time!
For reinstall App on B device in that situation yes, works, but I have to import default data at first launch, and if I do , and after running sync, those will be duplicated, but I hope this PrimaryKey will solve it! I’ll write you soon!

**Comment by spotlessicode** [comment]
Hi, I grabbed this version using Cocoapods, this is ok. I check the QSCoreDataCahngeManager.m file, and logged out, whether primaryKey exists, just for controll it. But I1m not sure how can I implement this NSString protocol method. I have 4 Entities, so I thought the easiest way in the subclass, I tried with this, but I have no primaryKey at all. 
**in Achieves+CoreDataProperties.m:**

```
#import "Achieves+CoreDataProperties.h"
#import <SyncKit/QSPrimaryKey.h>
@implementation Achieves (CoreDataProperties)

+(NSFetchRequest<Achieves *> *)fetchRequest {
	return [[NSFetchRequest alloc] initWithEntityName:@"Achieves"];
}
+(nonnull NSString*)primaryKey{
    Achieves *achieve = [[Achieves alloc]init];
    [achieve willAccessValueForKey:@"primaryKey"];
    NSString *key = [achieve primitiveValueForKey:@"primaryKey"];
    [achieve didAccessValueForKey:@"primaryKey"];
    return key;
}
@dynamic primaryKey;
@dynamic dateInstalled;
```


I tried just in **Achieves+CoreDataClass.m** file do the same, but I always get No PrimaryKey. I log out this here in **QSCoreDataCahngeManager.m** `- (void)performInitialSetup` :
```

 for (NSEntityDescription *entityDescription in entities) {
            NSLog(@"for loop entities");
            NSError *error = nil;
            NSString *primaryKey = [self identifierFieldNameForEntityOfType:entityDescription.managedObjectClassName];
            NSArray *objectIDs;
            
if (primaryKey) {
                //Get object identifiers using primary key
                NSLog(@"primaryKey!");
                objectIDs = [self.targetContext executeFetchRequestWithEntityName:entityDescription.name predicate:nil fetchLimit:0 resultType:NSDictionaryResultType propertiesToFetch:@[primaryKey] error:&error];
                objectIDs = [objectIDs valueForKey:primaryKey];
                NSLog(@"objectIDs primaryKey= %@", objectIDs);
} else {
                NSLog(@"NO primaryKey!");
                //Get objectIDs
                objectIDs = [self.targetContext executeFetchRequestWithEntityName:entityDescription.name predicate:nil fetchLimit:0 resultType:NSManagedObjectIDResultType error:&error];
                NSMutableArray *objectIDStrings = [NSMutableArray array];
                [objectIDs enumerateObjectsUsingBlock:^(NSManagedObjectID *  _Nonnull objectID, NSUInteger idx, BOOL * _Nonnull stop) {
                    [objectIDStrings addObject:[objectID.URIRepresentation absoluteString]];
                }];
                objectIDs = [objectIDStrings copy];
            }
```

**Comment by mentrena** [comment]
Hi,

I'm not sure I understand your case, but if you want to avoid duplicate data then don't you have a way to determine that two objects are the same? Normally you'd have a unique field that can be used to find duplicates, and in the case of SyncKit we'll need to get an object using a fetch request, so this unique field should be a stored property.

For example, if your class is called Achieves:

- Add a string property to it in your model, name it whatever you like. Let's say you name it `identifier`.
- Make your class conform to the protocol, normally done in Achieves+CoreDataClass.h:
```objc
#import <SyncKit/QSPrimaryKey.h>
....
@interface Achieves: NSManagedObject <QSPrimaryKey>
```

- Implement the protocol method in Achieves+CoreDataClass.m:
```objc
...
@implementation Achieves

+ (NSString *)primaryKey
{
    return @"identifier";
}

@end
```
- Make sure each instance of your class has a unique identifier when it's created:
```objc
....
Achieves *achieve = [NSEntityDescription insertNewObjectForEntityForName:@"Achieves" inManagedObjectContext:context];
achieve.identifier = @"some unique identifier"; //Maybe [[NSUUID UUID] UUIDString] ?
```

That way, if you device A creates an instance of Achieves with some identifier, and device B creates another instance of Achieves with that same identifier, when synchronization takes place they will be considered the same object so you won't get a duplicate.

It might help to take a look at QSCompany2 and QSEmployee2 in the example project.

**Comment by spotlessicode** [comment]
Hi,

I was confused, I thought I have to get the identifier's String values, not the identifier itself as a String. That was the problem :) If it returns the dynamic property as a String, everything works fine! It's great! 
I have made some tests, at the first launch the '**setupPrimaryKeysLookup**' is running, after that the '**performInitialSetup**'  and there are primaryKeys, and **I could list the objectIDs primaryKeys** in the log!
**I tested with these steps:** 
1. install App on device, launch first time, with default data import (15 objects)- 
2. Run synchronizer - check CloudKit create records
3. Delete the App from the device and reinstall it 
4. Run the App, run new sync, and first check: the objects on the device are not duplicated!
5. finally check the Records number at the Entity in CloudKit, there are just the default data import record's number(15). Earlier it was at second install 30, after that 45 ... 

**So it works like a charm!** Thanks for this clarification! I'm going to make some other tests - with more devices, and Lightweight Migration too. 

I'm going to **write a tutorial** today or tomorrow with SyncKit and with this objectidentifier deduplication method!
Have you any other advice for CloudKit recommended settings?, or other warning as the many-to-many relationship, something else? I'll share here the tutorial, if you find it useful, you can link it in Readme if you want. And later I'll need a data/results sharing function with other users too, so it would be great, if you could later further develop this SyncKit with public records too. ;) And I'll notice you about our Application as soon as it will be available in AppStore.

**Comment by mentrena** [comment]
Glad to hear it's working :)

> Have you any other advice for CloudKit recommended settings?

Just to be careful with your model versioning. Even if SyncKit can keep track of your objects after a lightweight migration you might still get errors if you have different devices running different versions of your app with different versions of your Core Data model.
Let's say you add property `name` to object `Person` in a new model version, and assign a value to it using the latest version of your app. If the user still has a previous version of your app in another device and you try to synchronize, SyncKit will try to assign to `person.name` using key-value coding and the app will crash because that model still doesn't have property `name`

I'm planning to add more support  for model versioning in future versions.

> if you could later further develop this SyncKit with public records too

I'd like to do that too but synchronization relies on `CKFetchRecordChangesOperation` and that only works on the private database, so that would take a different approach.

**Comment by spotlessicode** [comment]
I did the Tutorial, but I have a Problem with this PrimaryKey. I think there is a bug with CustomZoneID which always the same. 
[Here you can grab the Test tutorial Project](https://github.com/spotlessicode/SyncKitTest_CloudKit_CoreData_IOS10)!
The Problem is, that I had to did some changes with the ObjectModel. After that I tried Sync again, but with PriamryKEy(after delete Record Types in CloudKit, and delete App from devide), I have got Sync Error. I tried with new Bundle ID, and therefore new CloudKit Container, and new Device, new iCloud Account, but the Sync said again, the same Error: 
`<CKError 0x17004fd80: "Invalid Arguments" (12); "You can't save the same record twice: <CKRecordID: 0x170030080; Companies.FirstCompany:(QSCloudKitCustomZoneName:__defaultOwner__)>">`
**But if I have new Container, new device, new iCloud account, then it should  work!** If I have 2 users, and They are create the same Companies, if the A User is already created, then the B User couldn't create it with the same name?  There are 5 Companies, if the 6. Company name is "MyCompany" at A user, and if it has already been saved when user B type as his 6. Company the same "MyCompany" name, then B user will get Sync error, because in the Container there is already a "MyCompany" object, but that object is for the A User not for the B. 
You can test the project just click the Sync button and check the Log. I've already written the Tutorial with codes and screenshots, just at the end with this test... so I hope you can figured it out. Maybe this is with Custom Zones, [so there is a solution](http://stackoverflow.com/questions/37277962/error-saving-record-ckrecordid-to-server-invalid-attempt-to-update-record-from). I didn't check it deeply, but if you use always the same ZoneID, that is the Problem. 
**in QSCloudKitSynchronizer+CoreData.m"** 
`static NSString * const QSCloudKitCustomZoneName = @"QSCloudKitCustomZoneName";`

If we can solve this, I'm going to publish the tutorial with Steps, screenshots and code snippets. You will see, **the process:**
1. check First launch, or already launch, if first start import CoreData Defaults
2. With buttons add other Records to other Record Types
3. With button Start Sync process - with iCloudAccount Signed in and iCloud Enabled checks
4. During Snyc I log the companies count, so we can see if it duplicates(5 to 10 without PrimaryKey), and now, if you run it with different user new device, new CloudKit container, and it shows 5 companies instead of 0.

**Comment by mentrena** [comment]
I'll take a look at your code, but synchronization uses the private database so there's no way the data from two different users would conflict on iCloud.
Let me take a look and get back to you.

**Comment by mentrena** [comment]
```objc
Companies *c1= [NSEntityDescription insertNewObjectForEntityForName:@"Companies" inManagedObjectContext:self.managedObjectContext];
c1.cName = @"FirstCompany";
c1.cAddress = @"MilkyWay1";
    
Companies *c2= [NSEntityDescription insertNewObjectForEntityForName:@"Companies" inManagedObjectContext:self.managedObjectContext];
c2.cName = @"FirstCompany";
c2.cAddress = @"MilkyWay1";
```

You're assigning the same name to all your Companies objects, but cName was supposed to be your primaryKey, so it must have a unique value for each object. This would be causing problems when trying to upload CKRecords, since the CKRecordID is dependent on the object's unique value for its primary key.

On a sidenote, you should probably use `performBlockAndWait` to save your managed object context when `changeManagerRequestsContextSave:completion:` and `changeManager:didImportChanges:completion:` are called (and I see I need to update my example project to do the same)

**Comment by spotlessicode** [comment]
Hi, the Tutorial Post available here (with 1 device deduplication test & 2 device sync test): [http://www.spotlessicode.com/blog/posts/synckit-tutorial-cloudkit-with-coredata-ios10-obj-c ](http://www.spotlessicode.com/blog/posts/synckit-tutorial-cloudkit-with-coredata-ios10-obj-c )
The changes are (companies data and performBlock) on Git. 
And I did a little video about the Steps and the test part of the Tutorial (It will be embed in the Post).

**Comment by mentrena** [comment]
Cheers.
Published 0.3.1 with these fixes.
